Control.cpp

```diff
@@ -45,61 +50,284 @@ double Control::getAngleError(const double current_angle_x, const double desire)
+  if (bno_->getAngleY() < -kLimitInfBumper1 && bno_->getAngleY() > -kLimitSupBumper1) {
+    if (bno_->getAngleZ() > kRangeAngleZ || bno_->getAngleZ() < -kRangeAngleZ) {
+      return true; // It's a bumper level 1.
+    }
+    else {
+      return false; // It's a ramp.
+    }
+  }
+  else {
+    return false; // It's another bumper level.
+  }
 }
 
 bool Control::bumperLevel2() {
-  return (bno_.getAngleY() < -kLimitSupBumper1 
-  && bno_.getAngleY() > -kLimitSupBumper2 
-  && bno_.getAngleZ() > kRangeAngleZ 
-  || bno_.getAngleZ() < -kRangeAngleZ);
+  if(bno_->getAngleY() < -kLimitSupBumper1 && bno_->getAngleY() > -kLimitSupBumper2) {
+    if(bno_->getAngleZ() > kRangeAngleZ || bno_->getAngleZ() < -kRangeAngleZ) {
+      return true; // It's a bumper level 2.
+    }
+    else {
+    return false; // It's a ramp.
+    }
+  }
+  else {
+  return false; // It's another bumper level.
+  }
 }
 
 bool Control::bumperLevel3() {
-  return (bno_.getAngleY() < -kLimitSupBumper2 
-  && bno_.getAngleY() > -kLimitSupBumper3 
-  && bno_.getAngleZ() > kRangeAngleZ 
-  || bno_.getAngleZ() < -kRangeAngleZ);
+  if(bno_->getAngleY() < -kLimitSupBumper2 && bno_->getAngleY() > -kLimitSupBumper3) {
+    if (bno_->getAngleZ() > kRangeAngleZ || bno_->getAngleZ() < -kRangeAngleZ) {
+      return true; // It's a bumper level 3.
+    }
+    return false; //It's a ramp.
+  } 
+  return false; // It's another bumper level.
+}
+
+double Control::getPwmBNOLeft(const double desire) { // Positive.
+    double pwm_left = 0;
+    double pwm_right = 0;
+    double pwm_left_final = 0;
+    double pwm_right_final = 0;
+    double pwm_right_BNO = 0;
+    double pwm_left_BNO = 0;
+    double current_angle_x = bno_->getAngleX(); 
+    const double errorBNO = getAngleError(current_angle_x, desire);
+
+    if (errorBNO > 0) {
+        pwm_right = Common::kLimit_inf_pwm;
+        pwm_left_BNO = kPAdvance * errorBNO;
+        pwm_left_final = pwm_left_BNO + getPwmUltrasonicLeft();
+        }
+    else {
+        pwm_left = Common::kLimit_inf_pwm;
+        pwm_right_BNO = kPAdvance * errorBNO;
+        pwm_right_final = pwm_right_BNO + getPwmUltrasonicRight();;
+    }
+    return pwm_left_final;
+}
+
+double Control::getPwmBNORight(const double desire) { // Negative.
+    double pwm_left = 0;
+    double pwm_right = 0;
+    double pwm_left_final = 0;
+    double pwm_right_final = 0;
+    double pwm_right_BNO = 0;
+    double pwm_left_BNO = 0;
+    double current_angle_x = bno_->getAngleX(); 
+    const double errorBNO = getAngleError(current_angle_x, desire);
+
+    if (errorBNO > 0) {
+        pwm_right = Common::kLimit_inf_pwm;
+        pwm_left_BNO = kPAdvance * errorBNO;
+        pwm_left_final = pwm_left_BNO + getPwmUltrasonicLeft();
+        }
+    else {
+        pwm_left = Common::kLimit_inf_pwm;
+        pwm_right_BNO = kPAdvance * errorBNO;
+        pwm_right_final = pwm_right_BNO + getPwmUltrasonicRight();;  
+    }
+    return pwm_right_final;
+}
```

Comment:
These two can be one function; the code is almost exactly the same.

Reply:
Same with the ultrasonic ones.

Reply:
Can you help with this functions? I tried to put them together, but I had problem, because I return two values, the right and left pwm, I did not know how to return two values ​​in the same function